rfc/rfc-91/rfc-91.md

```diff
@@ -0,0 +1,145 @@
+<!--
+  Licensed to the Apache Software Foundation (ASF) under one or more
+  contributor license agreements.  See the NOTICE file distributed with
+  this work for additional information regarding copyright ownership.
+  The ASF licenses this file to You under the Apache License, Version 2.0
+  (the "License"); you may not use this file except in compliance with
+  the License.  You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+-->
+# RFC-91: Storage-based lock provider using conditional writes
+
+## Proposers
+
+- @alexr17
+
+## Approvers
+
+ - @yihua
+ - @danny0405
+
+## Status
+
+JIRA: [HUDI-9122](https://issues.apache.org/jira/browse/HUDI-9122)
+
+## Abstract
+
+Currently in Hudi, distributed locking relies on external systems like Zookeeper, which add complexity and extra dependencies. This RFC introduces a storage-based implementation of the `LockProvider` interface that utilizes conditional writes in cloud storage platforms (such as GCS and AWS S3) to implement a native distributed locking mechanism for Hudi. By directly integrating lock management with cloud storage, this solution reduces operational overhead, and ensures robust coordination during concurrent writes.
+
+## Background
+
+AWS S3 recently introduced conditional writes, and GCS and Azure storage already support them. This RFC leverages these features to implement a distributed lock provider for Hudi using a leader election algorithm. In this approach, each process attempts an atomic conditional write to a file calculated using the table base path. The first process to succeed is elected leader and takes charge of exclusive operations. This method provides a straightforward, reliable locking mechanism without the need for external lock providers.
+
+## Implementation
+
+This design implements a leader election algorithm for Apache Hudi using a single lock file per table stored in cloud storage. Each table’s lock is represented by a JSON file with the following fields:
```

Comment:
So the design mainly focuses on cloud storage, how about the HDFS which also has a large user base.

Reply:
Wow, looks like they just added java support: https://github.com/apache/opendal/pull/5664

So this might be doable. I think it still needs to be a pluggable implementation, in case OpenDAL does not support conditional writes in a storage type that someone wants to add a lock provider for.

Reply:
> I think it still needs to be a pluggable implementation

+1 as of now.

Reply:
For HDFS and OpenDAL we will track them here:
- https://issues.apache.org/jira/browse/HUDI-9161
- https://issues.apache.org/jira/browse/HUDI-9162

Since both of these implementations are not available yet we will proceed with current S3/GCS client library approach to unblock Flink for 1.0.2 release, but can hopefully refactor in the future. cc @yihua